base/array.jl

```diff
@@ -1805,15 +1805,19 @@ julia> filter(isodd, a)
 """
 filter(f, As::AbstractArray) = As[map(f, As)::AbstractArray{Bool}]
 
-function filter!(f, a::Vector)
-    insrt = 1
+function filter!(f, a::AbstractVector)
+    idx = eachindex(a)
+    state = start(idx)
+    (i, state) = next(idx, state)
```

Comment:
@nalimilan The `next` behaviour was kind of intentional here as we want `i` to be at `state+1` passed the end of the new last index when we get to the `deleteat!`. I suppose a `done(idx, state)` or `isempty(a)` check would avoid needing to call `deleteat!`, but we still run into the issue of going passed the end of the array with the `next(idx, state)` call in the for loop.